middle_end/flambda/invariant_params.ml

```diff
@@ -253,7 +264,8 @@ let analyse_functions ~backend ~param_to_param
                   ~callee_arg:(Parameter.var param) !relation)
          params)
     decls.funs;
-  transitive_closure !relation
+  if too_many_functions decls then set_to_top !relation
```

Comment:
This test skips the work of calling `transitive_closure`, but there are already a few traversals before that -- I count three traversals of `decls.funs`, some of which appear to perform non-trivial computation. Is there an easy, correct way to also skip that buildup work in the case where there are too many functions?

Reply:
Good question. I think I will need to change the mli, but maybe it isn't too bad.

Reply:
It turns out that all functions exposed in the mli can safely return empty sets or maps. Only `invariant_params_sources` is used in a context where it might make sense to know immediately if it is empty (in `unbox_specialised_args.ml`), but I think that if it was empty we would never have specialised the functions so we don't need to worry about that.
So I've changed `analyse_functions` to return an option, and adapted the callers appropriately.